# Notebook FOUR


* **Overview of this Notebook**

In this notebook we will explore and clean the `keywords.csv` dataset.

> **Load and Clean dataset**

Now we load the *keywords.csv* as `dataframe` and display top 5 records of the data

In [1]:
# Import Libraries
import pandas as pd
import numpy as np

pd.set_option("display.max_colwidth", None)

In [2]:
dataframe = pd.read_csv("C:/Users/sibak/My Drive/PROJECTS/DATA CLEANING - MOVIES DATASET/Data Cleaning - Movies Dataset/the-movies-dataset/keywords.csv")
dataframe.head(5)

id                                           keywords
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

##### Do some exploration of the dataset

In [3]:
dataframe.info() # display the info of the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46419 entries, 0 to 46418
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        46419 non-null  int64 
 1   keywords  46419 non-null  object
dtypes: int64(1), object(1)
memory usage: 725.4+ KB


In [4]:
len(dataframe['id'].unique()) # display the unique value to check how many they are

45432

above code shows that the unique of ID is `45432`, and in the previous code we saw that the number of rows/records are `46419`, where we expected to both numbers to be equal but they not so thus it could mean there are duplicates, or something else. 

In [5]:
dataframe.iloc[0].keywords # sample look into the keywords

"[{'id': 931, 'name': 'jealousy'}, {'id': 4290, 'name': 'toy'}, {'id': 5202, 'name': 'boy'}, {'id': 6054, 'name': 'friendship'}, {'id': 9713, 'name': 'friends'}, {'id': 9823, 'name': 'rivalry'}, {'id': 165503, 'name': 'boy next door'}, {'id': 170722, 'name': 'new toy'}, {'id': 187065, 'name': 'toy comes to life'}]"

In [6]:
dataframe.tail(5) # display the tail(bottom 5) of the dataset

id                                           keywords
46414  439050             [{'id': 10703, 'name': 'tragic love'}]
46415  111109  [{'id': 2679, 'name': 'artist'}, {'id': 14531,...
46416   67758                                                 []
46417  227506                                                 []
46418  461257                                                 []

###### Replace the empty columns with the word empty

We find that when we count the strings in the empty rows, the results show that the row is not empty. therefore we have to find another approach into separating these rows alone so that we cam manipulate them.

In [9]:
dataframe['string count']=dataframe.astype(str).apply(lambda x: x.str.len()).sum(1) # count the number of string in a record

In [10]:
dataframe # display the dataframe

id                                           keywords  string count
0         862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...           317
1        8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...           242
2       15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...           159
3       31357  [{'id': 818, 'name': 'based on novel'}, {'id':...           207
4       11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...           347
...       ...                                                ...           ...
46414  439050             [{'id': 10703, 'name': 'tragic love'}]            44
46415  111109  [{'id': 2679, 'name': 'artist'}, {'id': 14531,...           102
46416   67758                                                 []             7
46417  227506                                                 []             8
46418  461257                                                 []             8

[46419 rows x 3 columns]

In [11]:
dataframe['number of keywords'] = dataframe.astype(str).sum(axis=1).str.count("'name':") # count the number of keywords

In [12]:
dataframe

id                                           keywords  \
0         862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...   
1        8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...   
2       15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...   
3       31357  [{'id': 818, 'name': 'based on novel'}, {'id':...   
4       11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...   
...       ...                                                ...   
46414  439050             [{'id': 10703, 'name': 'tragic love'}]   
46415  111109  [{'id': 2679, 'name': 'artist'}, {'id': 14531,...   
46416   67758                                                 []   
46417  227506                                                 []   
46418  461257                                                 []   

       string count  number of keywords  
0               317                   9  
1               242                   6  
2               159                   4  
3               207                   5  
4               347                   9  
...             ...                 ...  
46414            44                   1  
46415           102                   3  
46416             7                   0  
46417             8                   0  
46418             8                   0  

[46419 rows x 4 columns]

In [14]:
dataframe[dataframe['number of keywords'] == 0] # display rows where number of keywords are equal to 0

id keywords  string count  number of keywords
7       45325       []             7                   0
32      78802       []             7                   0
36     139405       []             8                   0
50     117164       []             8                   0
55     124057       []             8                   0
...       ...      ...           ...                 ...
46412  222848       []             8                   0
46413   30840       []             7                   0
46416   67758       []             7                   0
46417  227506       []             8                   0
46418  461257       []             8                   0

[14795 rows x 4 columns]

In [17]:
dataframe.loc[dataframe['number of keywords'] == 0, 'keywords'] = "empty" # Replace the empty rows with word 'empty'

In [18]:
dataframe

id                                           keywords  \
0         862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...   
1        8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...   
2       15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...   
3       31357  [{'id': 818, 'name': 'based on novel'}, {'id':...   
4       11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...   
...       ...                                                ...   
46414  439050             [{'id': 10703, 'name': 'tragic love'}]   
46415  111109  [{'id': 2679, 'name': 'artist'}, {'id': 14531,...   
46416   67758                                              empty   
46417  227506                                              empty   
46418  461257                                              empty   

       string count  number of keywords  
0               317                   9  
1               242                   6  
2               159                   4  
3               207                   5  
4               347                   9  
...             ...                 ...  
46414            44                   1  
46415           102                   3  
46416             7                   0  
46417             8                   0  
46418             8                   0  

[46419 rows x 4 columns]

###### Clean the keywords column

I took this approach to clean the column(keywords):

1. assuming that all keywords are in this format `{'id': 165503, 'name': 'boy next door'}`.
2. I separated the dataframe of where number of keywords are greater than 1, so that i can clean the dataset and append the keywords only, and with its respective tmbid.
3. since the remaining dataframe is of where number of keywords are equal to 1 and 0, for 0 wont wont do any manipuilation. But for where is equal to 1, we will just remove the unncessary symbols.
4. at the end combine all this datarames into ONE dataframe

In [51]:
df_wkw = dataframe[dataframe['number of keywords'] > 1][['id','keywords']] # display with atleast 1 keyword

In [52]:
df = df_wkw # separate it into a diiferent variable for manipulation
df

id  \
0         862   
1        8844   
2       15602   
3       31357   
4       11862   
...       ...   
46404  106807   
46406  404604   
46409   84419   
46411  289923   
46415  111109   

                                                                                                                                                                                                                                                                                                                                                     keywords  
0                                  [{'id': 931, 'name': 'jealousy'}, {'id': 4290, 'name': 'toy'}, {'id': 5202, 'name': 'boy'}, {'id': 6054, 'name': 'friendship'}, {'id': 9713, 'name': 'friends'}, {'id': 9823, 'name': 'rivalry'}, {'id': 165503, 'name': 'boy next door'}, {'id': 170722, 'name': 'new toy'}, {'id': 187065, 'name': 'toy comes to life'}]  
1                                                                                                              [{'id': 10090, 'name': 'board game'}, {'id': 10941, 'name': 'disappearance'}, {'id': 15101, 'name': "based on children's book"}, {'id': 33467, 'name': 'new home'}, {'id': 158086, 'name': 'recluse'}, {'id': 158091, 'name': 'giant insect'}]  
2                                                                                                                                                                                                  [{'id': 1495, 'name': 'fishing'}, {'id': 12392, 'name': 'best friend'}, {'id': 179431, 'name': 'duringcreditsstinger'}, {'id': 208510, 'name': 'old men'}]  
3                                                                                                                                                  [{'id': 818, 'name': 'based on novel'}, {'id': 10131, 'name': 'interracial relationship'}, {'id': 14768, 'name': 'single mother'}, {'id': 15160, 'name': 'divorce'}, {'id': 33455, 'name': 'chick flick'}]  
4      [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'name': 'midlife crisis'}, {'id': 2246, 'name': 'confidence'}, {'id': 4995, 'name': 'aging'}, {'id': 5600, 'name': 'daughter'}, {'id': 10707, 'name': 'mother daughter relationship'}, {'id': 13149, 'name': 'pregnancy'}, {'id': 33358, 'name': 'contraception'}, {'id': 170521, 'name': 'gynecologist'}]  
...                                                                                                                                                                                                                                                                                                                                                       ...  
46404                                                                                                                                                                           [{'id': 5970, 'name': 'wrestling'}, {'id': 6075, 'name': 'sport'}, {'id': 154802, 'name': 'silent film'}, {'id': 214549, 'name': 'short'}, {'id': 230616, 'name': 'fat man'}]  
46406                                                                                                                                                                                                                                                                    [{'id': 155794, 'name': 'physical abuse'}, {'id': 190327, 'name': 'sexual assault'}]  
46409                                                                                        [{'id': 9748, 'name': 'revenge'}, {'id': 9826, 'name': 'murder'}, {'id': 10714, 'name': 'serial killer'}, {'id': 14512, 'name': 'new york city'}, {'id': 173245, 'name': 'sculptor'}, {'id': 179083, 'name': 'art critic'}, {'id': 189379, 'name': 'deformity'}]  
46411                                                                                                                                                                        [{'id': 616, 'name': 'witch'}, {'id': 2035, 'name': 'mythology'}, {'id': 3754, 'name': 'legend'}, {'id': 10714, '

In [53]:
df = df.reset_index(drop=True) # reset index numbers

In [54]:
df

id  \
0         862   
1        8844   
2       15602   
3       31357   
4       11862   
...       ...   
24996  106807   
24997  404604   
24998   84419   
24999  289923   
25000  111109   

                                                                                                                                                                                                                                                                                                                                                     keywords  
0                                  [{'id': 931, 'name': 'jealousy'}, {'id': 4290, 'name': 'toy'}, {'id': 5202, 'name': 'boy'}, {'id': 6054, 'name': 'friendship'}, {'id': 9713, 'name': 'friends'}, {'id': 9823, 'name': 'rivalry'}, {'id': 165503, 'name': 'boy next door'}, {'id': 170722, 'name': 'new toy'}, {'id': 187065, 'name': 'toy comes to life'}]  
1                                                                                                              [{'id': 10090, 'name': 'board game'}, {'id': 10941, 'name': 'disappearance'}, {'id': 15101, 'name': "based on children's book"}, {'id': 33467, 'name': 'new home'}, {'id': 158086, 'name': 'recluse'}, {'id': 158091, 'name': 'giant insect'}]  
2                                                                                                                                                                                                  [{'id': 1495, 'name': 'fishing'}, {'id': 12392, 'name': 'best friend'}, {'id': 179431, 'name': 'duringcreditsstinger'}, {'id': 208510, 'name': 'old men'}]  
3                                                                                                                                                  [{'id': 818, 'name': 'based on novel'}, {'id': 10131, 'name': 'interracial relationship'}, {'id': 14768, 'name': 'single mother'}, {'id': 15160, 'name': 'divorce'}, {'id': 33455, 'name': 'chick flick'}]  
4      [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'name': 'midlife crisis'}, {'id': 2246, 'name': 'confidence'}, {'id': 4995, 'name': 'aging'}, {'id': 5600, 'name': 'daughter'}, {'id': 10707, 'name': 'mother daughter relationship'}, {'id': 13149, 'name': 'pregnancy'}, {'id': 33358, 'name': 'contraception'}, {'id': 170521, 'name': 'gynecologist'}]  
...                                                                                                                                                                                                                                                                                                                                                       ...  
24996                                                                                                                                                                           [{'id': 5970, 'name': 'wrestling'}, {'id': 6075, 'name': 'sport'}, {'id': 154802, 'name': 'silent film'}, {'id': 214549, 'name': 'short'}, {'id': 230616, 'name': 'fat man'}]  
24997                                                                                                                                                                                                                                                                    [{'id': 155794, 'name': 'physical abuse'}, {'id': 190327, 'name': 'sexual assault'}]  
24998                                                                                        [{'id': 9748, 'name': 'revenge'}, {'id': 9826, 'name': 'murder'}, {'id': 10714, 'name': 'serial killer'}, {'id': 14512, 'name': 'new york city'}, {'id': 173245, 'name': 'sculptor'}, {'id': 179083, 'name': 'art critic'}, {'id': 189379, 'name': 'deformity'}]  
24999                                                                                                                                                                        [{'id': 616, 'name': 'witch'}, {'id': 2035, 'name': 'mythology'}, {'id': 3754, 'name': 'legend'}, {'id': 10714, '

###### Start the manipulation on the few first record, the define a function to manipulate the remaining dataset

In [25]:
# Separate the keywords in the first row
keywords_data = pd.Series(dataframe.iloc[0].keywords).str.split("}, {", expand=True).T

In [26]:
keywords_data

0
0              [{'id': 931, 'name': 'jealousy'
1                    'id': 4290, 'name': 'toy'
2                    'id': 5202, 'name': 'boy'
3             'id': 6054, 'name': 'friendship'
4                'id': 9713, 'name': 'friends'
5                'id': 9823, 'name': 'rivalry'
6        'id': 165503, 'name': 'boy next door'
7              'id': 170722, 'name': 'new toy'
8  'id': 187065, 'name': 'toy comes to life'}]

Separate again the keywords, and be left with only [name: keyword]

In [27]:
newkeywords = pd.Series(keywords_data.iloc[0]).str.split(", '", expand=True)

for i in range(1,len(keywords_data)):
    new_keywords_data_holder = pd.Series(keywords_data.iloc[i]).str.split(", '", expand=True)
    newkeywords = newkeywords.append(new_keywords_data_holder)
    
del newkeywords[0]

In [28]:
newkeywords

1
0             name': 'jealousy'
0                  name': 'toy'
0                  name': 'boy'
0           name': 'friendship'
0              name': 'friends'
0              name': 'rivalry'
0        name': 'boy next door'
0              name': 'new toy'
0  name': 'toy comes to life'}]

In [29]:
# Replace the [name] with [name_(number of th keyword)]

for x in range(len(newkeywords)):
    newkeywords.iloc[x] = newkeywords.iloc[x].str.replace("name':" , f"name_{x}: ")

In [30]:
newkeywords

1
0             name_0:  'jealousy'
0                  name_1:  'toy'
0                  name_2:  'boy'
0           name_3:  'friendship'
0              name_4:  'friends'
0              name_5:  'rivalry'
0        name_6:  'boy next door'
0              name_7:  'new toy'
0  name_8:  'toy comes to life'}]

In [31]:
# split them and make the [name_(number of th keyword) the index

newkeywords_split_df = pd.Series(newkeywords.iloc[0]).str.split(":", expand=True)

for i in range(1,len(keywords_data)):
    newkeywords_split = pd.Series(newkeywords.iloc[i]).str.split(":", expand=True)
    newkeywords_split_df = newkeywords_split_df.append(newkeywords_split)
    
newkeywords_split_df = newkeywords_split_df.set_index(0,inplace=False)

In [32]:
newkeywords_split_df

1
0                              
name_0               'jealousy'
name_1                    'toy'
name_2                    'boy'
name_3             'friendship'
name_4                'friends'
name_5                'rivalry'
name_6          'boy next door'
name_7                'new toy'
name_8    'toy comes to life'}]

In [33]:
newkeywords_Transposed = newkeywords_split_df.T # transpose the data

In [34]:
newkeywords_Transposed

0        name_0   name_1   name_2          name_3       name_4       name_5  \
1    'jealousy'    'toy'    'boy'    'friendship'    'friends'    'rivalry'   

0             name_6       name_7                   name_8  
1    'boy next door'    'new toy'    'toy comes to life'}]

In [35]:
# Combine all keywords into a single record value

Data = newkeywords_Transposed['name_0']

for i in range(1,len(newkeywords_Transposed.columns)):
    Data = Data + ' ' + newkeywords_Transposed[f'name_{i}']
      
edf = pd.DataFrame(Data)

In [36]:
edf

0
1    'jealousy'   'toy'   'boy'   'friendship'   ...

In [37]:
edf['tmbId'] = dataframe.iloc[0].id # add the tmbid to the record

In [38]:
edf

0  tmbId
1    'jealousy'   'toy'   'boy'   'friendship'   ...    862

###### define funtions that will manipulate the whole dataset

In [56]:
df_KW = pd.DataFrame()
    
keywords = df
    
for w in range(0,len(keywords)): # len(keywords)
    kw_data = pd.Series(keywords.iloc[w].keywords).str.split("}, {", expand=True).T
    
    nkw = pd.Series(kw_data.iloc[0]).str.split(", '", expand=True)
    for i in range(1,len(kw_data)):
        nkw_data_holder = pd.Series(kw_data.iloc[i]).str.split(", '", expand=True)
        nkw = nkw.append(nkw_data_holder)
    
    del nkw[0]
    
    for x in range(len(nkw)):
        nkw.iloc[x] = nkw.iloc[x].str.replace("name':" , f"name_{x}: ")
        
    nkw_split_df = pd.Series(nkw.iloc[0]).str.split(":", expand=True)
    
    for y in range(1,len(nkw)):
        nkw_split = pd.Series(nkw.iloc[y]).str.split(":", expand=True)
        nkw_split_df = nkw_split_df.append(nkw_split)
        
    nkw_split_df = nkw_split_df.set_index(0,inplace=False)
    
    nkw_T = nkw_split_df.T
    
    New_df = nkw_T['name_0']
    
    if len(nkw_T.columns) > 1:
        for k in range(1,len(nkw_T.columns)):
            New_df = New_df + ' + ' + nkw_T[f'name_{k}']
        
        df_kw = pd.DataFrame(New_df)
        
        df_kw['tmbId'] = keywords.iloc[w].id
        
        df_KW = df_KW.append(df_kw)
    
    else:
        df_kw = pd.DataFrame(New_df)
        
        df_kw['tmbId'] = keywords.iloc[w].id
        
        df_KW = df_KW.append(df_kw)

In [57]:
df_KW # display results

0  \
1                                 'jealousy' +   'toy' +   'boy' +   'friendship' +   'friends' +   'rivalry' +   'boy next door' +   'new toy' +   'toy comes to life'}]   
1                                                       'board game' +   'disappearance' +   "based on children's book" +   'new home' +   'recluse' +   'giant insect'}]   
1                                                                                                  'fishing' +   'best friend' +   'duringcreditsstinger' +   'old men'}]   
1                                                                   'based on novel' +   'interracial relationship' +   'single mother' +   'divorce' +   'chick flick'}]   
1     'baby' +   'midlife crisis' +   'confidence' +   'aging' +   'daughter' +   'mother daughter relationship' +   'pregnancy' +   'contraception' +   'gynecologist'}]   
..                                                                                                                                                                    ...   
1                                                                                                   'wrestling' +   'sport' +   'silent film' +   'short' +   'fat man'}]   
1                                                                                                                                 'physical abuse' +   'sexual assault'}]   
1                                                        'revenge' +   'murder' +   'serial killer' +   'new york city' +   'sculptor' +   'art critic' +   'deformity'}]   
1                                                                                           'witch' +   'mythology' +   'legend' +   'serial killer' +   'mockumentary'}]   
1                                                                                                                                       'artist' +   'play' +   'pinoy'}]   

     tmbId  
1      862  
1     8844  
1    15602  
1    31357  
1    11862  
..     ...  
1   106807  
1   404604  
1    84419  
1   289923  
1   111109  

[25003 rows x 2 columns]

In [62]:
df_KW.duplicated().any() # check for duplicates

1

In [63]:
df_KW.drop_duplicates() # drop duplicates

0  \
1                                 'jealousy' +   'toy' +   'boy' +   'friendship' +   'friends' +   'rivalry' +   'boy next door' +   'new toy' +   'toy comes to life'}]   
1                                                       'board game' +   'disappearance' +   "based on children's book" +   'new home' +   'recluse' +   'giant insect'}]   
1                                                                                                  'fishing' +   'best friend' +   'duringcreditsstinger' +   'old men'}]   
1                                                                   'based on novel' +   'interracial relationship' +   'single mother' +   'divorce' +   'chick flick'}]   
1     'baby' +   'midlife crisis' +   'confidence' +   'aging' +   'daughter' +   'mother daughter relationship' +   'pregnancy' +   'contraception' +   'gynecologist'}]   
..                                                                                                                                                                    ...   
1                                                                                                   'wrestling' +   'sport' +   'silent film' +   'short' +   'fat man'}]   
1                                                                                                                                 'physical abuse' +   'sexual assault'}]   
1                                                        'revenge' +   'murder' +   'serial killer' +   'new york city' +   'sculptor' +   'art critic' +   'deformity'}]   
1                                                                                           'witch' +   'mythology' +   'legend' +   'serial killer' +   'mockumentary'}]   
1                                                                                                                                       'artist' +   'play' +   'pinoy'}]   

     tmbId  
1      862  
1     8844  
1    15602  
1    31357  
1    11862  
..     ...  
1   106807  
1   404604  
1    84419  
1   289923  
1   111109  

[24627 rows x 2 columns]

In [ ]:
df_KW['number of keywords'] = df_KW.astype(str).sum(axis=1).str.count("'name':") # count the number of keywords

We now deal with the dataframe where number of keywords equal to 1

In [71]:
dataframe[dataframe['number of keywords'] == 1] # display with less 1 keyword

id                                     keywords  string count  \
39      34615       [{'id': 3644, 'name': 'south africa'}]            43   
52      49133  [{'id': 10183, 'name': 'independent film'}]            48   
71      28387            [{'id': 3616, 'name': 'college'}]            38   
74      36929             [{'id': 10624, 'name': 'bully'}]            37   
92       9283         [{'id': 155808, 'name': 'hometown'}]            40   
...       ...                                          ...           ...   
46392   84710     [{'id': 208422, 'name': 'paula dupree'}]            45   
46394   14008  [{'id': 171803, 'name': 'military school'}]            48   
46395   44330      [{'id': 169590, 'name': 'police raid'}]            44   
46410  390959      [{'id': 224180, 'name': 'blair witch'}]            45   
46414  439050       [{'id': 10703, 'name': 'tragic love'}]            44   

       number of keywords  
39                      1  
52                      1  
71                      1  
74                      1  
92                      1  
...                   ...  
46392                   1  
46394                   1  
46395                   1  
46410                   1  
46414                   1  

[6623 rows x 4 columns]

In [81]:
split_df = dataframe[dataframe['number of keywords'] == 1]

In [82]:
split_df.keywords.str.split(" 'name':",expand=True) # split the data

0                      1
39       [{'id': 3644,       'south africa'}]
52      [{'id': 10183,   'independent film'}]
71       [{'id': 3616,            'college'}]
74      [{'id': 10624,              'bully'}]
92     [{'id': 155808,           'hometown'}]
...                ...                    ...
46392  [{'id': 208422,       'paula dupree'}]
46394  [{'id': 171803,    'military school'}]
46395  [{'id': 169590,        'police raid'}]
46410  [{'id': 224180,        'blair witch'}]
46414   [{'id': 10703,        'tragic love'}]

[6623 rows x 2 columns]

In [83]:
split_df[['delete','keywords']] = split_df.keywords.str.split(" 'name':",expand=True)

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\frame.py:3641: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


In [84]:
split_df

id               keywords  string count  number of keywords  \
39      34615       'south africa'}]            43                   1   
52      49133   'independent film'}]            48                   1   
71      28387            'college'}]            38                   1   
74      36929              'bully'}]            37                   1   
92       9283           'hometown'}]            40                   1   
...       ...                    ...           ...                 ...   
46392   84710       'paula dupree'}]            45                   1   
46394   14008    'military school'}]            48                   1   
46395   44330        'police raid'}]            44                   1   
46410  390959        'blair witch'}]            45                   1   
46414  439050        'tragic love'}]            44                   1   

                delete  
39       [{'id': 3644,  
52      [{'id': 10183,  
71       [{'id': 3616,  
74      [{'id': 10624,  
92     [{'id': 155808,  
...                ...  
46392  [{'id': 208422,  
46394  [{'id': 171803,  
46395  [{'id': 169590,  
46410  [{'id': 224180,  
46414   [{'id': 10703,  

[6623 rows x 5 columns]

In [85]:
# delete the unneccessary columns
del split_df['delete']
del split_df['string count']

In [86]:
split_df

id               keywords  number of keywords
39      34615       'south africa'}]                   1
52      49133   'independent film'}]                   1
71      28387            'college'}]                   1
74      36929              'bully'}]                   1
92       9283           'hometown'}]                   1
...       ...                    ...                 ...
46392   84710       'paula dupree'}]                   1
46394   14008    'military school'}]                   1
46395   44330        'police raid'}]                   1
46410  390959        'blair witch'}]                   1
46414  439050        'tragic love'}]                   1

[6623 rows x 3 columns]

We now check where number of keywords are equal to 0

In [87]:
dataframe[dataframe['number of keywords'] == 0]

id keywords  string count  number of keywords
7       45325    empty             7                   0
32      78802    empty             7                   0
36     139405    empty             8                   0
50     117164    empty             8                   0
55     124057    empty             8                   0
...       ...      ...           ...                 ...
46412  222848    empty             8                   0
46413   30840    empty             7                   0
46416   67758    empty             7                   0
46417  227506    empty             8                   0
46418  461257    empty             8                   0

[14795 rows x 4 columns]

In [88]:
df0 = dataframe[dataframe['number of keywords'] == 0]

In [89]:
del df0['string count'] # delete string count column

In [90]:
df0

id keywords  number of keywords
7       45325    empty                   0
32      78802    empty                   0
36     139405    empty                   0
50     117164    empty                   0
55     124057    empty                   0
...       ...      ...                 ...
46412  222848    empty                   0
46413   30840    empty                   0
46416   67758    empty                   0
46417  227506    empty                   0
46418  461257    empty                   0

[14795 rows x 3 columns]

In [91]:
df_KW

0  \
1                                 'jealousy' +   'toy' +   'boy' +   'friendship' +   'friends' +   'rivalry' +   'boy next door' +   'new toy' +   'toy comes to life'}]   
1                                                       'board game' +   'disappearance' +   "based on children's book" +   'new home' +   'recluse' +   'giant insect'}]   
1                                                                                                  'fishing' +   'best friend' +   'duringcreditsstinger' +   'old men'}]   
1                                                                   'based on novel' +   'interracial relationship' +   'single mother' +   'divorce' +   'chick flick'}]   
1     'baby' +   'midlife crisis' +   'confidence' +   'aging' +   'daughter' +   'mother daughter relationship' +   'pregnancy' +   'contraception' +   'gynecologist'}]   
..                                                                                                                                                                    ...   
1                                                                                                   'wrestling' +   'sport' +   'silent film' +   'short' +   'fat man'}]   
1                                                                                                                                 'physical abuse' +   'sexual assault'}]   
1                                                        'revenge' +   'murder' +   'serial killer' +   'new york city' +   'sculptor' +   'art critic' +   'deformity'}]   
1                                                                                           'witch' +   'mythology' +   'legend' +   'serial killer' +   'mockumentary'}]   
1                                                                                                                                       'artist' +   'play' +   'pinoy'}]   

     tmbId  
1      862  
1     8844  
1    15602  
1    31357  
1    11862  
..     ...  
1   106807  
1   404604  
1    84419  
1   289923  
1   111109  

[25003 rows x 2 columns]

In [95]:
df_KW_new = df_KW.reset_index(drop=True)

In [96]:
df_KW_new

0  \
0                                    'jealousy' +   'toy' +   'boy' +   'friendship' +   'friends' +   'rivalry' +   'boy next door' +   'new toy' +   'toy comes to life'}]   
1                                                          'board game' +   'disappearance' +   "based on children's book" +   'new home' +   'recluse' +   'giant insect'}]   
2                                                                                                     'fishing' +   'best friend' +   'duringcreditsstinger' +   'old men'}]   
3                                                                      'based on novel' +   'interracial relationship' +   'single mother' +   'divorce' +   'chick flick'}]   
4        'baby' +   'midlife crisis' +   'confidence' +   'aging' +   'daughter' +   'mother daughter relationship' +   'pregnancy' +   'contraception' +   'gynecologist'}]   
...                                                                                                                                                                      ...   
24998                                                                                                  'wrestling' +   'sport' +   'silent film' +   'short' +   'fat man'}]   
24999                                                                                                                                'physical abuse' +   'sexual assault'}]   
25000                                                       'revenge' +   'murder' +   'serial killer' +   'new york city' +   'sculptor' +   'art critic' +   'deformity'}]   
25001                                                                                          'witch' +   'mythology' +   'legend' +   'serial killer' +   'mockumentary'}]   
25002                                                                                                                                      'artist' +   'play' +   'pinoy'}]   

        tmbId  
0         862  
1        8844  
2       15602  
3       31357  
4       11862  
...       ...  
24998  106807  
24999  404604  
25000   84419  
25001  289923  
25002  111109  

[25003 rows x 2 columns]

In [98]:
df_KW_new.astype(str).sum(axis=1).str.count(" + ")

0        9
1        6
2        4
3        5
4        9
        ..
24998    5
24999    2
25000    7
25001    5
25002    3
Length: 25003, dtype: int64

In [99]:
df_KW_new['number of keywords'] = df_KW_new.astype(str).sum(axis=1).str.count(" + ") # count number of keywords

In [100]:
df_KW_new

0  \
0                                    'jealousy' +   'toy' +   'boy' +   'friendship' +   'friends' +   'rivalry' +   'boy next door' +   'new toy' +   'toy comes to life'}]   
1                                                          'board game' +   'disappearance' +   "based on children's book" +   'new home' +   'recluse' +   'giant insect'}]   
2                                                                                                     'fishing' +   'best friend' +   'duringcreditsstinger' +   'old men'}]   
3                                                                      'based on novel' +   'interracial relationship' +   'single mother' +   'divorce' +   'chick flick'}]   
4        'baby' +   'midlife crisis' +   'confidence' +   'aging' +   'daughter' +   'mother daughter relationship' +   'pregnancy' +   'contraception' +   'gynecologist'}]   
...                                                                                                                                                                      ...   
24998                                                                                                  'wrestling' +   'sport' +   'silent film' +   'short' +   'fat man'}]   
24999                                                                                                                                'physical abuse' +   'sexual assault'}]   
25000                                                       'revenge' +   'murder' +   'serial killer' +   'new york city' +   'sculptor' +   'art critic' +   'deformity'}]   
25001                                                                                          'witch' +   'mythology' +   'legend' +   'serial killer' +   'mockumentary'}]   
25002                                                                                                                                      'artist' +   'play' +   'pinoy'}]   

        tmbId  number of keywords  
0         862                   9  
1        8844                   6  
2       15602                   4  
3       31357                   5  
4       11862                   9  
...       ...                 ...  
24998  106807                   5  
24999  404604                   2  
25000   84419                   7  
25001  289923                   5  
25002  111109                   3  

[25003 rows x 3 columns]

###### Rename the dataframes so that we can concatenate them together

In [108]:
column_rename={0:'keywords'}

df_KW_new_c_renamed = df_KW_new.rename(columns=column_rename)

In [109]:
df_KW_new_c_renamed

keywords  \
0                                    'jealousy' +   'toy' +   'boy' +   'friendship' +   'friends' +   'rivalry' +   'boy next door' +   'new toy' +   'toy comes to life'}]   
1                                                          'board game' +   'disappearance' +   "based on children's book" +   'new home' +   'recluse' +   'giant insect'}]   
2                                                                                                     'fishing' +   'best friend' +   'duringcreditsstinger' +   'old men'}]   
3                                                                      'based on novel' +   'interracial relationship' +   'single mother' +   'divorce' +   'chick flick'}]   
4        'baby' +   'midlife crisis' +   'confidence' +   'aging' +   'daughter' +   'mother daughter relationship' +   'pregnancy' +   'contraception' +   'gynecologist'}]   
...                                                                                                                                                                      ...   
24998                                                                                                  'wrestling' +   'sport' +   'silent film' +   'short' +   'fat man'}]   
24999                                                                                                                                'physical abuse' +   'sexual assault'}]   
25000                                                       'revenge' +   'murder' +   'serial killer' +   'new york city' +   'sculptor' +   'art critic' +   'deformity'}]   
25001                                                                                          'witch' +   'mythology' +   'legend' +   'serial killer' +   'mockumentary'}]   
25002                                                                                                                                      'artist' +   'play' +   'pinoy'}]   

        tmbId  number of keywords  
0         862                   9  
1        8844                   6  
2       15602                   4  
3       31357                   5  
4       11862                   9  
...       ...                 ...  
24998  106807                   5  
24999  404604                   2  
25000   84419                   7  
25001  289923                   5  
25002  111109                   3  

[25003 rows x 3 columns]

In [110]:
column_rename={'id':'tmbId'}

df0_c_renamed = df0.rename(columns=column_rename)

In [112]:
df0_c_renamed

tmbId keywords  number of keywords
7       45325    empty                   0
32      78802    empty                   0
36     139405    empty                   0
50     117164    empty                   0
55     124057    empty                   0
...       ...      ...                 ...
46412  222848    empty                   0
46413   30840    empty                   0
46416   67758    empty                   0
46417  227506    empty                   0
46418  461257    empty                   0

[14795 rows x 3 columns]

In [113]:
column_rename={'id':'tmbId'}

split_df_c_renamed = split_df.rename(columns=column_rename)

In [114]:
split_df_c_renamed

tmbId               keywords  number of keywords
39      34615       'south africa'}]                   1
52      49133   'independent film'}]                   1
71      28387            'college'}]                   1
74      36929              'bully'}]                   1
92       9283           'hometown'}]                   1
...       ...                    ...                 ...
46392   84710       'paula dupree'}]                   1
46394   14008    'military school'}]                   1
46395   44330        'police raid'}]                   1
46410  390959        'blair witch'}]                   1
46414  439050        'tragic love'}]                   1

[6623 rows x 3 columns]

##### Concatenate the dataframes

In [116]:
final_df = pd.concat([df_KW_new_c_renamed, split_df_c_renamed, df0_c_renamed])

In [117]:
final_df

keywords  \
0                                    'jealousy' +   'toy' +   'boy' +   'friendship' +   'friends' +   'rivalry' +   'boy next door' +   'new toy' +   'toy comes to life'}]   
1                                                          'board game' +   'disappearance' +   "based on children's book" +   'new home' +   'recluse' +   'giant insect'}]   
2                                                                                                     'fishing' +   'best friend' +   'duringcreditsstinger' +   'old men'}]   
3                                                                      'based on novel' +   'interracial relationship' +   'single mother' +   'divorce' +   'chick flick'}]   
4        'baby' +   'midlife crisis' +   'confidence' +   'aging' +   'daughter' +   'mother daughter relationship' +   'pregnancy' +   'contraception' +   'gynecologist'}]   
...                                                                                                                                                                      ...   
46412                                                                                                                                                                  empty   
46413                                                                                                                                                                  empty   
46416                                                                                                                                                                  empty   
46417                                                                                                                                                                  empty   
46418                                                                                                                                                                  empty   

        tmbId  number of keywords  
0         862                   9  
1        8844                   6  
2       15602                   4  
3       31357                   5  
4       11862                   9  
...       ...                 ...  
46412  222848                   0  
46413   30840                   0  
46416   67758                   0  
46417  227506                   0  
46418  461257                   0  

[46421 rows x 3 columns]

In [118]:
final_df['keywords']=final_df['keywords'].str.replace("}]", '') # remove symbols

C:\Users\sibak\AppData\Local\Temp/ipykernel_24432/1464099055.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  final_df['keywords']=final_df['keywords'].str.replace("}]", '')


In [119]:
final_df

keywords  \
0                                    'jealousy' +   'toy' +   'boy' +   'friendship' +   'friends' +   'rivalry' +   'boy next door' +   'new toy' +   'toy comes to life'   
1                                                          'board game' +   'disappearance' +   "based on children's book" +   'new home' +   'recluse' +   'giant insect'   
2                                                                                                     'fishing' +   'best friend' +   'duringcreditsstinger' +   'old men'   
3                                                                      'based on novel' +   'interracial relationship' +   'single mother' +   'divorce' +   'chick flick'   
4        'baby' +   'midlife crisis' +   'confidence' +   'aging' +   'daughter' +   'mother daughter relationship' +   'pregnancy' +   'contraception' +   'gynecologist'   
...                                                                                                                                                                    ...   
46412                                                                                                                                                                empty   
46413                                                                                                                                                                empty   
46416                                                                                                                                                                empty   
46417                                                                                                                                                                empty   
46418                                                                                                                                                                empty   

        tmbId  number of keywords  
0         862                   9  
1        8844                   6  
2       15602                   4  
3       31357                   5  
4       11862                   9  
...       ...                 ...  
46412  222848                   0  
46413   30840                   0  
46416   67758                   0  
46417  227506                   0  
46418  461257                   0  

[46421 rows x 3 columns]

In [120]:
# Save the DataSet
final_df.to_csv('Movies_keywords_NB4.csv')